### ISSUE TITLE
AUC calculation is probably wrong

### ISSUE THREAD

**Issue by RLstat** [catboost/catboost#335]
Problem: AUC number 
catboost version: catboost 0.8.1.1
Operating System: amazon linux
CPU: {CPU}

I've recently used catboost for classification problems. The training starts normal, but after a few iterations, the training and validation metric ('AUC') value suddenly dropped and then back up to 1( which means perfect fit and is impossible). 

when I use the model to predict new data(predict_proba), all entries returns NA.

![talking_data_catboost_issue](https://user-images.githubusercontent.com/18621301/39666940-033e7404-507a-11e8-8ebe-21dcd011ccfa.png)

```python

    cb_params = {

    'iterations': 50, 
    'thread_count': 6, 
    'loss_function': 'Logloss',
    'verbose' : True,
    'random_seed' : 42,
    'save_snapshot' : True,
    'snapshot_file' : os.path.join(output_path, 'snapshot'),

    'l2_leaf_reg' : 10,
    'scale_pos_weight' : 200, 
    'one_hot_max_size' : 30,
    'max_ctr_complexity' : 3,
    'leaf_estimation_iterations' : 8,
    'depth': 6,
    'rsm':0.9,
    'learning_rate' : 0.1,
    'eval_metric' : 'AUC'
    }
    
    categorical_features_indices = [i for i, f in enumerate(predictors) 
                                     if f in categorical_features] 
    
    cb_train = cb.Pool(dtrain[predictors].values, label=dtrain[target].values,
                     feature_names=predictors,
                     cat_features=categorical_features_indices
                    )
    
    cb_valid = cb.Pool(dvalid[predictors].values, label=dvalid[target].values,
                     feature_names=predictors,
                     cat_features=categorical_features_indices)
    
    del dtrain
    del dvalid
    gc.collect()
    gc.collect()
    
    cb_model = cb.CatBoostClassifier(**cb_params)
    
    cb_model.fit(cb_train, 
                 eval_set=cb_valid, 
                 use_best_model = True)
    
    train_auc = cb_model.eval_metrics(cb_train, metrics = ['AUC'], 
                                      ntree_start = cb_model.tree_count_ - 1, 
                                      ntree_end=cb_model.tree_count_ , 
                                      eval_period=1 
                                     )
    
    print(train_auc)
    
    val_auc = cb_model.eval_metrics(cb_valid,  metrics = ['AUC'], 
                                      ntree_start = cb_model.tree_count_ - 1, 
                                      ntree_end=cb_model.tree_count_ , 
                                      eval_period=1 
                                     )
    
    val_predict = cb_model.predict_proba(cb_valid, ntree_start=cb_model.tree_count_ - 1, 
                                    ntree_end=cb_model.tree_count_)
    
    
    print(val_auc)
    
    print("best iteration is: {}".format(cb_model.tree_count_))
    print("train AUC is {}".format(train_auc))
    print("validation AUC is {}".format(val_auc))

```

**Comment by annaveronika** [comment]
The training has diverged, we'll look into auc, but for now you can try increasing L2 regularisation or changing optimization method from newton to gradient:
l2_leaf_reg=100 or leaf_estimation_method='Gradient'

**Comment by annaveronika** [comment]
@RLstat We are working on improving the cases when training diverges. If it is possible to share your dataset, it would help us a lot and it would improve the quality for your dataset.
